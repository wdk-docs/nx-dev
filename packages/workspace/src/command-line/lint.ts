import {
  createProjectGraphAsync,
  pruneExternalNodes,
} from '../core/project-graph';
import { WorkspaceIntegrityChecks } from './workspace-integrity-checks';
import { FileData, workspaceLayout } from '../core/file-utils';
import { output } from '../utilities/output';
import * as path from 'path';

export async function workspaceLint(): Promise<void> {
  const graph = await createProjectGraphAsync();
  const allWorkspaceFiles = graph.allWorkspaceFiles;
  const projectGraph = pruneExternalNodes(graph);

  const cliErrorOutputConfigs = new WorkspaceIntegrityChecks(
    projectGraph,
    readAllFilesFromAppsAndLibs(allWorkspaceFiles)
  ).run();

  if (cliErrorOutputConfigs.length > 0) {
    cliErrorOutputConfigs.forEach((errorConfig) => {
      output.error(errorConfig);
    });
    process.exit(1);
  }
}

function readAllFilesFromAppsAndLibs(allWorkspaceFiles: FileData[]) {
  const wl = workspaceLayout();
  return allWorkspaceFiles
    .map((f) => f.file)
    .filter(
      (f) => f.startsWith(`${wl.appsDir}/`) || f.startsWith(`${wl.libsDir}/`)
    )
    .filter((f) => !path.basename(f).startsWith('.'));
}
