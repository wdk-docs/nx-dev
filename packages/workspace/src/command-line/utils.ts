import * as yargsParser from 'yargs-parser';
import * as yargs from 'yargs';
import { readNxJson } from '../core/file-utils';
import { output } from '../utilities/output';
import { names } from '@nrwl/devkit';
import type { NxAffectedConfig } from '@nrwl/devkit';

const runOne: string[] = [
  'target',
  'configuration',
  'prod',
  'runner',
  'parallel',
  'max-parallel',
  'exclude',
  'only-failed',
  'help',
  'with-deps',
  'skip-nx-cache',
  'scan',
  'hide-cached-output',
];

const runMany: string[] = [...runOne, 'projects', 'all'];

const runAffected: string[] = [
  ...runOne,
  'untracked',
  'uncommitted',
  'all',
  'base',
  'head',
  'files',
  'plain',
  'select',
];

export interface RawNxArgs extends NxArgs {
  prod?: boolean;
}

export interface NxArgs {
  target?: string;
  configuration?: string;
  runner?: string;
  parallel?: number;
  untracked?: boolean;
  uncommitted?: boolean;
  all?: boolean;
  base?: string;
  head?: string;
  exclude?: string[];
  files?: string[];
  onlyFailed?: boolean;
  'only-failed'?: boolean;
  verbose?: boolean;
  help?: boolean;
  version?: boolean;
  plain?: boolean;
  withDeps?: boolean;
  'with-deps'?: boolean;
  projects?: string[];
  select?: string;
  skipNxCache?: boolean;
  'skip-nx-cache'?: boolean;
  'hide-cached-output'?: boolean;
  hideCachedOutput?: boolean;
  scan?: boolean;
}

const ignoreArgs = ['$0', '_'];

export function splitArgsIntoNxArgsAndOverrides(
  args: yargs.Arguments,
  mode: 'run-one' | 'run-many' | 'affected' | 'print-affected',
  options = { printWarnings: true }
): { nxArgs: NxArgs; overrides: yargs.Arguments } {
  const nxSpecific =
    mode === 'run-one' ? runOne : mode === 'run-many' ? runMany : runAffected;

  const nxArgs: RawNxArgs = {};
  const overrides = yargsParser(args._ as string[], {
    configuration: {
      'strip-dashed': true,
      'dot-notation': false,
    },
  });
  // This removes the overrides from the nxArgs._
  args._ = overrides._;

  delete overrides._;

  Object.entries(args).forEach(([key, value]) => {
    const dasherized = names(key).fileName;
    if (nxSpecific.includes(dasherized) || dasherized.startsWith('nx-')) {
      if (value !== undefined) nxArgs[key] = value;
    } else if (!ignoreArgs.includes(key)) {
      overrides[key] = value;
    }
  });

  if (mode === 'run-many') {
    if (!nxArgs.projects) {
      nxArgs.projects = [];
    } else {
      nxArgs.projects = (args.projects as string)
        .split(',')
        .map((p: string) => p.trim());
    }
  }

  if (nxArgs.prod) {
    delete nxArgs.prod;
    nxArgs.configuration = 'production';
  }

  if (mode === 'affected') {
    if (options.printWarnings && nxArgs.all) {
      output.warn({
        title: `Running affected:* commands with --all can result in very slow builds.`,
        bodyLines: [
          `${output.bold(
            '--all'
          )} is not meant to be used for any sizable project or to be used in CI.`,
          '',
          `${output.colors.gray(
            'Learn more about checking only what is affected: '
          )}https://nx.dev/latest/angular/cli/affected#affected.`,
        ],
      });
    }

    if (
      !nxArgs.files &&
      !nxArgs.uncommitted &&
      !nxArgs.untracked &&
      !nxArgs.base &&
      !nxArgs.head &&
      !nxArgs.all &&
      args._.length >= 3
    ) {
      nxArgs.base = args._[1] as string;
      nxArgs.head = args._[2] as string;
    }

    // Allow setting base and head via environment variables (lower priority then direct command arguments)
    if (!nxArgs.base && process.env.NX_BASE) {
      nxArgs.base = process.env.NX_BASE;
      if (options.printWarnings) {
        output.note({
          title: `No explicit --base argument provided, but found environment variable NX_BASE so using its value as the affected base: ${output.bold(
            `${nxArgs.base}`
          )}`,
        });
      }
    }
    if (!nxArgs.head && process.env.NX_HEAD) {
      nxArgs.head = process.env.NX_HEAD;
      if (options.printWarnings) {
        output.note({
          title: `No explicit --head argument provided, but found environment variable NX_HEAD so using its value as the affected head: ${output.bold(
            `${nxArgs.head}`
          )}`,
        });
      }
    }

    if (!nxArgs.base) {
      const affectedConfig = getAffectedConfig();
      nxArgs.base = affectedConfig.defaultBase;

      // No user-provided arguments to set the affected criteria, so inform the user of the defaults being used
      if (
        options.printWarnings &&
        !nxArgs.head &&
        !nxArgs.files &&
        !nxArgs.uncommitted &&
        !nxArgs.untracked &&
        !nxArgs.all
      ) {
        output.note({
          title: `Affected criteria defaulted to --base=${output.bold(
            `${nxArgs.base}`
          )} --head=${output.bold('HEAD')}`,
        });
      }
    }
  }

  if (!nxArgs.skipNxCache) {
    nxArgs.skipNxCache = process.env.NX_SKIP_NX_CACHE === 'true';
  }

  if (args['parallel'] === 'false' || args['parallel'] === false) {
    nxArgs['parallel'] = 1;
  } else if (
    args['parallel'] === 'true' ||
    args['parallel'] === true ||
    args['parallel'] === ''
  ) {
    nxArgs['parallel'] = Number(
      nxArgs['maxParallel'] || nxArgs['max-parallel'] || 3
    );
  } else if (args['parallel'] !== undefined) {
    nxArgs['parallel'] = Number(args['parallel']);
  }

  return { nxArgs, overrides };
}

export function getAffectedConfig(): NxAffectedConfig {
  const config = readNxJson();

  return {
    defaultBase: config.affected?.defaultBase || 'main',
  };
}
