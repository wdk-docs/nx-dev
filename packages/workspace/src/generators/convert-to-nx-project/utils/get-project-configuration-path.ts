import { ProjectConfiguration } from '@nrwl/devkit';
import { join } from 'path';

export function getProjectConfigurationPath(
  configuration: ProjectConfiguration
) {
  return join(configuration.root, 'project.json');
}
