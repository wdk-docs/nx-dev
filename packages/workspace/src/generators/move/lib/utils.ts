import {
  getWorkspaceLayout,
  joinPathFragments,
  ProjectConfiguration,
  Tree,
} from '@nrwl/devkit';

import { Schema } from '../schema';

/**
 * This helper function ensures that we don't move libs or apps
 * outside of the folders they should be in.
 *
 * This will break if someone isn't using the default libs/apps
 * folders. In that case, they're on their own :/
 */
export function getDestination(
  host: Tree,
  schema: Schema,
  project: ProjectConfiguration
): string {
  const projectType = project.projectType;

  const workspaceLayout = getWorkspaceLayout(host);

  let rootFolder = workspaceLayout.libsDir;
  if (projectType === 'application') {
    rootFolder = workspaceLayout.appsDir;
  }
  return joinPathFragments(rootFolder, schema.destination);
}

/**
 * Replaces slashes with dashes
 *
 * @param path
 */
export function getNewProjectName(path: string): string {
  return path.replace(/\//g, '-');
}

/**
 * Normalizes slashes (removes duplicates)
 *
 * @param input
 */
export function normalizeSlashes(input: string): string {
  return input
    .split('/')
    .filter((x) => !!x)
    .join('/');
}
