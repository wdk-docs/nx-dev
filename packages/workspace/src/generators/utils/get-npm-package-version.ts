export function getNpmPackageVersion(packageName: string): string | null {
  try {
    const version = require('child_process').execSync(
      `npm view ${packageName} version`,
      { stdio: ['pipe', 'pipe', 'ignore'] }
    );

    if (version) {
      return version
        .toString()
        .trim()
        .replace(/^\n*|\n*$/g, '');
    }
  } catch (err) {}
  return null;
}
