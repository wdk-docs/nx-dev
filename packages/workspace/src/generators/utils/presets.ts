export enum Preset {
  Apps = 'apps',
  Empty = 'empty', // same as apps, deprecated
  Core = 'core',
  NPM = 'npm', // same as core, deprecated
  TS = 'ts',
  WebComponents = 'web-components',
  Angular = 'angular',
  AngularWithNest = 'angular-nest',
  React = 'react',
  ReactWithExpress = 'react-express',
  ReactNative = 'react-native',
  NextJs = 'next',
  Nest = 'nest',
  Express = 'express',
}
