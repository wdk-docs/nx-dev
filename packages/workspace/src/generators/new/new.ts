import {
  addDependenciesToPackageJson,
  convertNxGenerator,
  formatFiles,
  getPackageManagerCommand,
  getWorkspacePath as devkitGetWorkspacePath,
  installPackagesTask,
  names,
  PackageManager,
  Tree,
  updateJson,
} from '@nrwl/devkit';

import { join } from 'path';
import * as yargsParser from 'yargs-parser';
import { spawn, SpawnOptions } from 'child_process';

import { gte } from 'semver';

import { workspaceGenerator } from '../workspace/workspace';
import { nxVersion } from '../../utils/versions';
import { Preset } from '../utils/presets';
import {
  checkGitVersion,
  deduceDefaultBase,
} from '../../utilities/default-base';
import { getNpmPackageVersion } from '../utils/get-npm-package-version';

export interface Schema {
  cli: 'nx' | 'angular';
  directory: string;
  name: string;
  appName: string;
  npmScope?: string;
  skipInstall?: boolean;
  skipGit?: boolean;
  style?: string;
  nxCloud?: boolean;
  preset: string;
  commit?: { name: string; email: string; message?: string };
  defaultBase: string;
  linter: 'tslint' | 'eslint';
  packageManager?: PackageManager;
}

export interface NormalizedSchema extends Schema {
  presetVersion?: string;
}

function generatePreset(host: Tree, opts: NormalizedSchema) {
  const cliCommand = opts.cli === 'angular' ? 'ng' : 'nx';
  const parsedArgs = yargsParser(process.argv, {
    boolean: ['interactive'],
  });
  const spawnOptions = {
    stdio: [process.stdin, process.stdout, process.stderr],
    shell: true,
    cwd: join(host.root, opts.directory),
  };
  const pmc = getPackageManagerCommand();
  const executable = `${pmc.exec} ${cliCommand}`;
  const args = getPresetArgs(opts);

  return new Promise<void>((resolve, reject) => {
    spawn(executable, args, spawnOptions).on('close', (code: number) => {
      if (code === 0) {
        resolve();
      } else {
        const message = 'Workspace creation failed, see above.';
        reject(new Error(message));
      }
    });
  });

  function getPresetArgs(options: NormalizedSchema) {
    if (Object.values(Preset).some((val) => val === options.preset)) {
      // supported presets
      return getDefaultArgs(options);
    }
    return getThirdPartyPresetArgs();
  }

  function getDefaultArgs(opts: Schema) {
    return [
      `g`,
      `@nrwl/workspace:preset`,
      `--name=${opts.appName}`,
      opts.style ? `--style=${opts.style}` : null,
      opts.linter ? `--linter=${opts.linter}` : null,
      opts.npmScope ? `--npmScope=${opts.npmScope}` : `--npmScope=${opts.name}`,
      opts.preset ? `--preset=${opts.preset}` : null,
      opts.packageManager ? `--packageManager=${opts.packageManager}` : null,
      `--cli=${cliCommand}`,
      parsedArgs.interactive ? '--interactive=true' : '--interactive=false',
    ].filter((e) => !!e);
  }

  function getThirdPartyPresetArgs() {
    const thirdPartyPkgArgs = Object.entries(opts).reduce(
      (acc, [key, value]) => {
        if (value === true) {
          acc.push(`--${key}`);
        } else if (value === false) {
          acc.push(`--no-${key}`);
          // nxWorkspaceRoot breaks Tao CLI if incorrectly set, so need to exclude it.
          // TODO(jack): Should read in the preset schema and only pass the options specified.
        } else if (key !== 'nxWorkspaceRoot') {
          // string, number (don't handle arrays or nested objects)
          acc.push(`--${key}=${value}`);
        }
        return acc;
      },
      []
    );
    return [`g`, `${opts.preset}:preset`, ...thirdPartyPkgArgs];
  }
}

async function initializeGitRepo(
  host: Tree,
  rootDirectory: string,
  options: Schema
) {
  const execute = (args: ReadonlyArray<string>, ignoreErrorStream = false) => {
    const outputStream = 'ignore';
    const errorStream = ignoreErrorStream ? 'ignore' : process.stderr;
    const spawnOptions: SpawnOptions = {
      stdio: [process.stdin, outputStream, errorStream],
      shell: true,
      cwd: join(host.root, rootDirectory),
      env: {
        ...process.env,
        ...(options.commit.name
          ? {
              GIT_AUTHOR_NAME: options.commit.name,
              GIT_COMMITTER_NAME: options.commit.name,
            }
          : {}),
        ...(options.commit.email
          ? {
              GIT_AUTHOR_EMAIL: options.commit.email,
              GIT_COMMITTER_EMAIL: options.commit.email,
            }
          : {}),
      },
    };
    return new Promise<void>((resolve, reject) => {
      spawn('git', args, spawnOptions).on('close', (code) => {
        if (code === 0) {
          resolve();
        } else {
          reject(code);
        }
      });
    });
  };
  const gitVersion = checkGitVersion();
  if (!gitVersion) {
    return;
  }
  const insideRepo = await execute(
    ['rev-parse', '--is-inside-work-tree'],
    true
  ).then(
    () => true,
    () => false
  );
  if (insideRepo) {
    console.info(
      `Directory is already under version control. Skipping initialization of git.`
    );
    return;
  }
  const defaultBase = options.defaultBase || deduceDefaultBase();
  if (gte(gitVersion, '2.28.0')) {
    await execute(['init', '-b', defaultBase]);
  } else {
    await execute(['init']);
    await execute(['checkout', '-b', defaultBase]); // Git < 2.28 doesn't support -b on git init.
  }
  await execute(['add', '.']);
  if (options.commit) {
    const message = options.commit.message || 'initial commit';
    await execute(['commit', `-m "${message}"`]);
  }
  console.info('Successfully initialized git.');
}

export async function newGenerator(host: Tree, options: Schema) {
  if (
    options.skipInstall &&
    options.preset !== Preset.Apps &&
    options.preset !== Preset.Core &&
    options.preset !== Preset.TS &&
    options.preset !== Preset.Empty &&
    options.preset !== Preset.NPM
  ) {
    throw new Error(`Cannot select a preset when skipInstall is set to true.`);
  }
  if (options.skipInstall && options.nxCloud) {
    throw new Error(`Cannot select nxCloud when skipInstall is set to true.`);
  }

  if (devkitGetWorkspacePath(host)) {
    throw new Error(
      'Cannot generate a new workspace within an existing workspace'
    );
  }

  options = normalizeOptions(options);

  if (
    host.exists(options.name) &&
    !host.isFile(options.name) &&
    host.children(options.name).length > 0
  ) {
    throw new Error(
      `${join(host.root, options.name)} is not an empty directory.`
    );
  }

  await workspaceGenerator(host, { ...options, nxCloud: undefined } as any);

  setDefaultLinter(host, options);
  addPresetDependencies(host, options);
  addCloudDependencies(host, options);

  await formatFiles(host);
  return async () => {
    installPackagesTask(host, false, options.directory, options.packageManager);
    await generatePreset(host, options);
    if (!options.skipGit) {
      try {
        await initializeGitRepo(host, options.directory, options);
      } catch (e) {
        console.error(
          `Could not initialize git repository. Error: ${e.message}`
        );
      }
    }
  };
}

export default newGenerator;
export const newSchematic = convertNxGenerator(newGenerator);

function addCloudDependencies(host: Tree, options: Schema) {
  if (options.nxCloud) {
    return addDependenciesToPackageJson(
      host,
      {},
      { '@nrwl/nx-cloud': 'latest' },
      join(options.directory, 'package.json')
    );
  }
}

function getPresetDependencies(preset: string, version?: string) {
  switch (preset) {
    case Preset.TS:
      return { dependencies: {}, dev: { '@nrwl/js': nxVersion } };

    case Preset.Angular:
      return { dependencies: { '@nrwl/angular': nxVersion }, dev: {} };

    case Preset.AngularWithNest:
      return {
        dependencies: { '@nrwl/angular': nxVersion },
        dev: { '@nrwl/nest': nxVersion },
      };

    case Preset.Express:
      return { dependencies: {}, dev: { '@nrwl/express': nxVersion } };

    case Preset.Nest:
      return { dependencies: {}, dev: { '@nrwl/nest': nxVersion } };

    case Preset.NextJs:
      return { dependencies: { '@nrwl/next': nxVersion }, dev: {} };

    case Preset.React:
      return { dependencies: {}, dev: { '@nrwl/react': nxVersion } };

    case Preset.ReactWithExpress:
      return {
        dependencies: {},
        dev: {
          '@nrwl/react': nxVersion,
          '@nrwl/express': nxVersion,
        },
      };

    case Preset.ReactNative:
      return { dependencies: {}, dev: { '@nrwl/react-native': nxVersion } };

    case Preset.WebComponents:
      return { dependencies: {}, dev: { '@nrwl/web': nxVersion } };

    default: {
      return {
        dev: {},
        dependencies: { [preset]: version ?? getNpmPackageVersion(preset) },
      };
    }
  }
}

function addPresetDependencies(host: Tree, options: NormalizedSchema) {
  if (
    options.preset === Preset.Apps ||
    options.preset === Preset.Core ||
    options.preset === Preset.Empty ||
    options.preset === Preset.NPM
  ) {
    return;
  }
  const { dependencies, dev } = getPresetDependencies(
    options.preset,
    options.presetVersion
  );
  return addDependenciesToPackageJson(
    host,
    dependencies,
    dev,
    join(options.directory, 'package.json')
  );
}

function normalizeOptions(options: NormalizedSchema): NormalizedSchema {
  options.name = names(options.name).fileName;
  if (!options.directory) {
    options.directory = options.name;
  }

  // If the preset already contains a version in the name
  // -- my-package@2.0.1
  // -- @scope/package@version
  const match = options.preset.match(
    /^(?<package>(@.+\/)?[^@]+)(@(?<version>\d+\.\d+\.\d+))?$/
  );
  if (match) {
    options.preset = match.groups.package;
    options.presetVersion = match.groups.version;
  }

  return options;
}

function setDefaultLinter(host: Tree, options: Schema) {
  const { linter, preset } = options;
  // Don't do anything if someone doesn't pick angular
  if (preset !== Preset.Angular && preset !== Preset.AngularWithNest) {
    return;
  }

  switch (linter) {
    case 'eslint': {
      setESLintDefault(host, options);
      break;
    }
    case 'tslint': {
      setTSLintDefault(host, options);
      break;
    }
  }
}

/**
 * This sets ESLint as the default for any schematics that default to TSLint
 */
function setESLintDefault(host: Tree, options: Schema) {
  updateJson(host, getWorkspacePath(host, options), (json) => {
    setDefault(json, '@nrwl/angular', 'application', 'linter', 'eslint');
    setDefault(json, '@nrwl/angular', 'library', 'linter', 'eslint');
    setDefault(
      json,
      '@nrwl/angular',
      'storybook-configuration',
      'linter',
      'eslint'
    );
    return json;
  });
}

/**
 * This sets TSLint as the default for any schematics that default to ESLint
 */
function setTSLintDefault(host: Tree, options: Schema) {
  updateJson(host, getWorkspacePath(host, options), (json) => {
    setDefault(json, '@nrwl/workspace', 'library', 'linter', 'tslint');
    setDefault(json, '@nrwl/cypress', 'cypress-project', 'linter', 'tslint');
    setDefault(json, '@nrwl/cypress', 'cypress-project', 'linter', 'tslint');
    setDefault(json, '@nrwl/node', 'application', 'linter', 'tslint');
    setDefault(json, '@nrwl/node', 'library', 'linter', 'tslint');
    setDefault(json, '@nrwl/nest', 'application', 'linter', 'tslint');
    setDefault(json, '@nrwl/nest', 'library', 'linter', 'tslint');
    setDefault(json, '@nrwl/express', 'application', 'linter', 'tslint');
    setDefault(json, '@nrwl/express', 'library', 'linter', 'tslint');

    return json;
  });
}

function getWorkspacePath(host: Tree, { directory, cli }: Schema) {
  return join(directory, cli === 'angular' ? 'angular.json' : 'workspace.json');
}

function setDefault(
  json: any,
  collectionName: string,
  generatorName: string,
  key: string,
  value: any
) {
  if (!json.schematics) json.schematics = {};
  if (
    json.schematics[collectionName] &&
    json.schematics[collectionName][generatorName]
  ) {
    json.schematics[collectionName][generatorName][key] = value;
  } else if (json.schematics[`${collectionName}:${generatorName}`]) {
    json.schematics[`${collectionName}:${generatorName}`][key] = value;
  } else {
    json.schematics[collectionName] = json.schematics[collectionName] || {};
    json.schematics[collectionName][generatorName] = { [key]: value };
  }
}
