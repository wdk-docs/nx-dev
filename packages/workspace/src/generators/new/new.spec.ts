import { createTree } from '@nrwl/devkit/testing';
import { readJson, Tree, writeJson, PackageManager } from '@nrwl/devkit';
import { newGenerator, Schema } from './new';
import { Linter } from '../../utils/lint';
import { Preset } from '../utils/presets';

const defaultOptions: Omit<Schema, 'name' | 'directory' | 'appName'> = {
  cli: 'nx',
  preset: Preset.Apps,
  skipInstall: false,
  skipGit: false,
  linter: Linter.EsLint,
  defaultBase: 'main',
};

describe('new', () => {
  let tree: Tree;

  beforeEach(() => {
    tree = createTree();
  });

  it('should generate an empty workspace.json', async () => {
    await newGenerator(tree, {
      ...defaultOptions,
      name: 'my-workspace',
      directory: 'my-workspace',
      npmScope: 'npmScope',
      appName: 'app',
    });
    expect(readJson(tree, 'my-workspace/workspace.json')).toMatchSnapshot();
  });

  it('should generate an empty nx.json', async () => {
    await newGenerator(tree, {
      ...defaultOptions,
      name: 'my-workspace',
      directory: 'my-workspace',
      npmScope: 'npmScope',
      appName: 'app',
    });
    expect(readJson(tree, 'my-workspace/nx.json')).toMatchSnapshot();
  });

  describe('--preset', () => {
    describe.each([[Preset.Empty], [Preset.Angular], [Preset.React]])(
      '%s',
      (preset) => {
        it('should generate necessary npm dependencies', async () => {
          await newGenerator(tree, {
            ...defaultOptions,
            name: 'my-workspace',
            directory: 'my-workspace',
            npmScope: 'npmScope',
            appName: 'app',
            preset,
          });

          expect(readJson(tree, 'my-workspace/package.json')).toMatchSnapshot();
        });
      }
    );
  });

  describe('--packageManager', () => {
    describe.each([['npm'], ['yarn'], ['pnpm']])(
      '%s',
      (packageManager: PackageManager) => {
        it('should set the packageManager in nx.json', async () => {
          await newGenerator(tree, {
            ...defaultOptions,
            name: 'my-workspace',
            directory: 'my-workspace',
            npmScope: 'npmScope',
            appName: 'app',
            cli: 'angular',
            packageManager,
          });

          const nxJson = readJson(tree, 'my-workspace/nx.json');
          expect(nxJson.cli.packageManager).toEqual(packageManager);
        });
      }
    );
  });

  it('should not modify any existing files', async () => {
    const packageJson = {
      dependencies: {
        existing: 'latest',
      },
    };
    const eslintConfig = {
      rules: {},
    };
    writeJson(tree, 'package.json', packageJson);
    writeJson(tree, '.eslintrc.json', eslintConfig);

    await newGenerator(tree, {
      ...defaultOptions,
      name: 'my-workspace',
      directory: 'my-workspace',
      npmScope: 'npmScope',
      appName: 'app',
    });

    expect(readJson(tree, 'package.json')).toEqual(packageJson);
    expect(readJson(tree, '.eslintrc.json')).toEqual(eslintConfig);
  });

  it('should throw an error when the directory is not empty', async () => {
    tree.write('my-workspace/file.txt', '');

    try {
      await newGenerator(tree, {
        ...defaultOptions,
        name: 'my-workspace',
        directory: 'my-workspace',
        npmScope: 'npmScope',
        appName: 'app',
      });
      fail('Generating into a non-empty directory should error.');
    } catch (e) {}
  });
});
