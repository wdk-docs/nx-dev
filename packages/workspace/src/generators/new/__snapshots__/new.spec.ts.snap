// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`new --preset angular should generate necessary npm dependencies 1`] = `
Object {
  "dependencies": Object {
    "@nrwl/angular": "*",
  },
  "devDependencies": Object {
    "@nrwl/cli": "*",
    "@nrwl/tao": "*",
    "@nrwl/workspace": "*",
    "@types/node": "16.11.7",
    "prettier": "^2.5.1",
    "typescript": "~4.4.3",
  },
  "license": "MIT",
  "name": "my-workspace",
  "private": true,
  "scripts": Object {
    "build": "nx build",
    "start": "nx serve",
    "test": "nx test",
  },
  "version": "0.0.0",
}
`;

exports[`new --preset empty should generate necessary npm dependencies 1`] = `
Object {
  "dependencies": Object {},
  "devDependencies": Object {
    "@nrwl/cli": "*",
    "@nrwl/tao": "*",
    "@nrwl/workspace": "*",
    "@types/node": "16.11.7",
    "prettier": "^2.5.1",
    "typescript": "~4.4.3",
  },
  "license": "MIT",
  "name": "my-workspace",
  "private": true,
  "scripts": Object {
    "build": "nx build",
    "start": "nx serve",
    "test": "nx test",
  },
  "version": "0.0.0",
}
`;

exports[`new --preset react should generate necessary npm dependencies 1`] = `
Object {
  "dependencies": Object {},
  "devDependencies": Object {
    "@nrwl/cli": "*",
    "@nrwl/react": "*",
    "@nrwl/tao": "*",
    "@nrwl/workspace": "*",
    "@types/node": "16.11.7",
    "prettier": "^2.5.1",
    "typescript": "~4.4.3",
  },
  "license": "MIT",
  "name": "my-workspace",
  "private": true,
  "scripts": Object {
    "build": "nx build",
    "start": "nx serve",
    "test": "nx test",
  },
  "version": "0.0.0",
}
`;

exports[`new should generate an empty nx.json 1`] = `
Object {
  "affected": Object {
    "defaultBase": "main",
  },
  "cli": Object {
    "defaultCollection": "@nrwl/workspace",
  },
  "implicitDependencies": Object {
    ".eslintrc.json": "*",
    "package.json": Object {
      "dependencies": "*",
      "devDependencies": "*",
    },
  },
  "npmScope": "npmScope",
  "targetDependencies": Object {
    "build": Array [
      Object {
        "projects": "dependencies",
        "target": "build",
      },
    ],
  },
  "tasksRunnerOptions": Object {
    "default": Object {
      "options": Object {
        "cacheableOperations": Array [
          "build",
          "lint",
          "test",
          "e2e",
        ],
      },
      "runner": "@nrwl/workspace/tasks-runners/default",
    },
  },
}
`;

exports[`new should generate an empty workspace.json 1`] = `
Object {
  "projects": Object {},
  "version": 2,
}
`;
